sdks/python/apache_beam/transforms/window_test.py

```diff
@@ -252,6 +276,50 @@ def test_timestamped_with_combiners(self):
       assert_that(mean_per_window, equal_to([(0, 2.0), (1, 7.0)]),
                   label='assert:mean')
 
+  @attr('ValidatesRunner')
+  def test_custom_windows(self):
+    with TestPipeline() as p:
+      result = (p | Create([TimestampedValue(x, x * 100) for x in range(7)])
+                | 'custom window' >> WindowInto(TestCustomWindows(250, 200))
+                | 'insert key' >> Map(lambda v: ('key', v.value))
+                | GroupByKey())
+
+      assert_that(result, equal_to([('key', [0, 1, 2]),
+                                    ('key', [3, 4]),
+                                    ('key', [5]),
+                                    ('key', [6])]))
+
+  @attr('ValidatesRunner')
+  def test_windows_idempotency(self):
+    with TestPipeline() as p:
+      result = (p | Create([(x, x * 2) for x in range(5)])
```

Comment:
You should create timestamped values in Create instead of down below.

Reply:
I tried that but it failed.

Dug this issue more and found this is a known issue with Python timestamped value, see https://github.com/apache/beam/pull/8206. In short, to have the downstream ptransforms treat an timestamped element as an element with a timestamp instead of a timestamped value object, there must be a DoFn (could be a simple Map(lamda x: x)) applied to the timestamped elements. (This should be fixed, but probably not in this PR....)

I found this test class already have a helper fun timestamped_key_values, so I reuse it in the new tests.